New getting started guide (#3704)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/start-here/step-3.mdx

---
title: "Step 3 - Protected pages"
description: Learn how to use FusionAuth by using an example application that has a protected page defined.
section: get started
subcategory: start here
navOrder: 0
codeRoot: https://raw.githubusercontent.com/fusionauth/fusionauth-example-get-started/main
---

<div className="border border-green-700 px-3 rounded-md">
  <i className="fa-duotone fa-stopwatch fa-2x mr-2 text-green-700 align-middle"/> Estimated time to complete this step: **5 minutes**
</div>

## Protecting pages

> Now that we have the FusionAuth and the example app running, let's take a look at how the application protects pages that require a user to be logged in.

Reviewer: Could we make this more concrete and specific to this example? i.e. tell the reader we are going to show them how the Changebank app only allows a logged-in user to access the account page. This not only protects access to people's accounts, but it also tells the Changebank app whose account to load up. Or something like that.

Open the file `src/index.mts` and find the route named `/account`. It should look something like this:

```typescript
app.get("/account", async (req, res) => {
  if (!await sdk.userHasAccess(req, res, ["admin", "user"])) {
    res.redirect(302, '/login');
    return;
  }

  res.sendFile(path.join(__dirname, '../templates/account.html'));
});
```

The first check is where the application determines if the user is logged in or not. This is done by calling the SDK function `userHasAccess` (you can ignore the last parameter, which we will cover in the next section). This function returns true if the user has access, which implicitly verifies that they are logged in. Let's take a look at this function.

> Open the file `src/sdk.ts` and find the function named `userHasAccess`. One note is that this application app uses `jose` which is a Node library used for validating and parsing JWT. If you want to learn more about `jose`, you can visit the project at Github here: https://github.com/panva/jose

Reviewer: ```suggestion
Open the file `src/sdk.ts` and find the function named `userHasAccess`. One note is that this application uses `jose` which is a Node library used for validating and parsing JWTs. If you want to learn more about `jose`, you can visit the project at Github here: https://github.com/panva/jose
```

> The `userHasAccess` method begins by loading the user from the request. This is accomplished by calling the `getUser` function of the SDK. If the user comes back `null`, then this function returns false.

Reviewer: Copied the bit about the user not being logged in if the JWT comes back null from down below, thought it helped clarify.

```suggestion
The `userHasAccess` method begins by loading the user from the request. This is accomplished by calling the `getUser` function of the SDK, which returns the user's access token. If the access token comes back `null`, that means the user is not logged in and this function returns false.
```

```typescript
async userHasAccess(req: Request, res: Response, roles: Array<string>): Promise<boolean> {
  const jwt = await this.getUser(req, res);

  // ...
}
```

Let's take a quick peek at the `getUser` function of the SDK. This function looks like this:

```typescript

> let accessToken = req.cookies[this.configuration.accessTokenCookieName];

Reviewer: Maybe add comments that map code sections to the description below.
```suggestion
// 1. Load the user's access token from a cookie in the request
let accessToken = req.cookies[this.configuration.accessTokenCookieName];
```

if (!accessToken) {
  return null;
}

> let payload: JWTPayload;

Reviewer: // 2. Verify that the access token is valid
```suggestion
let payload: JWTPayload;
```

try {
  payload = (await jose.jwtVerify(accessToken, this.JWKS, {
    issuer: this.configuration.oauthIssuer,
    audience: this.configuration.applicationId,
  })).payload;
} catch (e) {
  // ...
}
```

> This function loads a cookie from the request that contains the user's access token. The access token is part of the OAuth specification and FusionAuth access tokens are also JWTs. The second part of this function verifies that the JWT is valid using `jose`'s `jwtVerify` function. This function validates the cryptographic signature of the JWT and also ensures that the JWT is not expired and all of the claims are valid.

Reviewer: ```suggestion
The first part of this function loads a cookie from the request that contains the user's access token. The access token is part of the OAuth specification, and is generated after the user successfully logs in. FusionAuth uses JWTs for access tokens. 

The second part of this function verifies that the JWT is valid using `jose`'s `jwtVerify` function. This function validates the cryptographic signature of the JWT and also ensures that the JWT is not expired and all of the claims are valid.
```

> We won't go into detail about JWTs and all of the claims here, but you can read our [OAuth token documentation](http://localhost:3000/docs/lifecycle/authenticate-users/oauth/tokens) to learn more.

Reviewer: It might be helpful to have and `Aside` that's like "What's in a JWT?" on this page for people who may not have come across them. It could just be a one-line description like, "A FusionAuth JWT holds verifiable information about a user. It's digitally signed to ensure that the contents are authentic. For more information... "

Open the file `src/sdk.ts` and find the function named `userHasAccess`. One note is that this application uses `jose` which is a Node library used for validating and parsing [JSON Web Tokens (JWTs)](/docs/lifecycle/authenticate-users/login-api/json-web-tokens). If you want to learn more about `jose`, you can visit the project at Github here: https://github.com/panva/jose

The `userHasAccess` method begins by loading the user from the request. This is accomplished by calling the `getUser` function of the SDK, which returns the user's access token. If the access token comes back `null`, that means the user is not logged in and this function returns false.

```typescript
async userHasAccess(req: Request, res: Response, roles: Array<string>): Promise<boolean> {
  const jwt = await this.getUser(req, res);

  // ...
}
```

Let's take a quick peek at the `getUser` function of the SDK. This function looks like this:

> ```typescript

Reviewer 2: I would prefer if we used RemoteCode to pull in these fragments (like here: https://github.com/FusionAuth/fusionauth-site/blob/main/astro/src/content/quickstarts/quickstart-ruby-rails-web.mdx?plain=1#L72 ) because it means the code will always be in sync with the example application repo. You can also use tags if you want to pull in fragments and you can also pull in a specific value from a JSON file including a kickstart file (more doco here: https://github.com/FusionAuth/fusionauth-astro-components )

// 1. Load the user's access token from a cookie in the request
let accessToken = req.cookies[this.configuration.accessTokenCookieName];
if (!accessToken) {
  return null;
}

// 2. Verify that the access token is valid
let payload: JWTPayload;
try {
  payload = (await jose.jwtVerify(accessToken, this.JWKS, {
    issuer: this.configuration.oauthIssuer,
    audience: this.configuration.applicationId,
  })).payload;
} catch (e) {
  // ...
}
```

The first part of this function loads a cookie from the request that contains the user's access token. The access token is part of the OAuth specification, and is generated after the user successfully logs in. FusionAuth uses JWTs for access tokens. 

> The second part of this function verifies that the JWT is valid using `jose`'s `jwtVerify` function. This function validates the cryptographic signature of the JWT and also ensures that the JWT is not expired and all of the claims are valid.

Reviewer 2: ```suggestion
The second part of this function verifies that the JWT is valid using `jose`'s `jwtVerify` function. This function validates the cryptographic signature of the JWT and also ensures that the JWT is not expired and the issuer and audience claims are expected values.
```

Author: I'm thinking about rejecting this change. We haven't described these claims and I think it will confuse the reader unless they already know about JWTs or have read our JWT doc.

> We won't go into detail about JWTs and all of the claims here, but you can read our [OAuth token documentation](http://localhost:3000/docs/lifecycle/authenticate-users/oauth/tokens) to learn more.

Reviewer 2: ```suggestion
We won't go into detail about JWTs and all of the claims here, but you can read our [OAuth token documentation](/docs/lifecycle/authenticate-users/oauth/tokens) to learn more.
```
